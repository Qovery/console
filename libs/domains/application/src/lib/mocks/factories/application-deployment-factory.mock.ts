import { DeploymentHistoryApplication, StateEnum } from 'qovery-typescript-axios'
import { Chance } from 'chance'

const chance = new Chance()

export const applicationDeploymentsFactoryMock = (howMany: number): DeploymentHistoryApplication[] =>
  Array.from({ length: howMany }).map((_, index) => ({
    id: index.toString(),
    created_at: new Date().toString(),
    updated_at: new Date().toString(),
    name: chance.name(),
    status: chance.pickone(Object.values([StateEnum.DEPLOYED, StateEnum.RUNNING, StateEnum.STOP_ERROR])),
  }))
