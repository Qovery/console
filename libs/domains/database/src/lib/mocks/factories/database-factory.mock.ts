import { Chance } from 'chance'
import { Database, DatabaseTypeEnum, DatabaseModeEnum, DatabaseAccessibilityEnum } from 'qovery-typescript-axios'

const chance = new Chance()

export const databaseFactoryMock = (howMany: number): Database[] =>
  Array.from({ length: howMany }).map((_, index) => ({
    id: `${index}`,
    created_at: new Date().toString(),
    updated_at: new Date().toString(),
    name: chance.name(),
    type: DatabaseTypeEnum.POSTGRESQL,
    version: '12',
    mode: DatabaseModeEnum.CONTAINER,
    accessibility: DatabaseAccessibilityEnum.PRIVATE,
    cpu: 1000,
    memory: 1024,
    storage: 1024,
    environment: {
      id: chance.guid(),
    },
    host: chance.name(),
    port: 80,
    maximum_cpu: 10,
    maximum_memory: 10,
    disk_encrypted: false,
  }))
