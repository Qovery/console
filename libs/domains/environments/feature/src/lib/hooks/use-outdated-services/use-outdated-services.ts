import { useQueries, useQuery } from '@tanstack/react-query'
import { type Commit } from 'qovery-typescript-axios'
import { P, match } from 'ts-pattern'
import { type Application, type Job, isApplication, isJob } from '@qovery/domains/services/data-access'
import { isJobGitSource } from '@qovery/shared/enums'
import { unionTypeGuard } from '@qovery/shared/util-types'
import { queries } from '@qovery/state/util-queries'

export interface UseOutdatedServicesProps {
  environmentId?: string
}

export type OutdatedService = (Application | Job) & { commits: Commit[] }

export function useOutdatedServices({ environmentId }: UseOutdatedServicesProps) {
  const { data: services = [], isLoading: isServicesLoading } = useQuery({
    ...queries.services.list(environmentId!),
    enabled: Boolean(environmentId),
  })

  const getGitRepository = (service: Application | Job) =>
    match(service)
      .with({ serviceType: 'APPLICATION' }, ({ git_repository }) => git_repository)
      .with({ serviceType: 'JOB', source: P.when(isJobGitSource) }, ({ source }) => source.docker?.git_repository)
      .otherwise(() => undefined)

  const gitServices = services
    .filter(unionTypeGuard([isApplication, isJob]))
    .filter((service) => Boolean(getGitRepository(service)))
  const queryResults = useQueries({
    queries: gitServices.map(({ id, serviceType }) => ({
      ...queries.services.listCommits({ serviceId: id, serviceType }),
    })),
  })

  const outdatedServices: OutdatedService[] = []

  for (let i = 0; i < queryResults.length; i++) {
    const { data: commits = [] } = queryResults[i]
    const service = gitServices[i]
    const gitRepository = getGitRepository(service)
    if (gitRepository?.deployed_commit_id !== commits[0]?.git_commit_id) {
      outdatedServices.push({
        ...service,
        commits,
      })
    }
  }

  return { data: outdatedServices, isLoading: queryResults.some(({ isLoading }) => isLoading) || isServicesLoading }
}

export default useOutdatedServices
