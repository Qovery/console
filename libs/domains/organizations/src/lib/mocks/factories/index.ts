export * from './organizations-factory.mock'
