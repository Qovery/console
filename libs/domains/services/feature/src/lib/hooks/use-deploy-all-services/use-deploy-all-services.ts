import { useMutation, useQueryClient } from '@tanstack/react-query'
import { useNavigate } from 'react-router-dom'
import { mutations } from '@qovery/domains/services/data-access'
import { ENVIRONMENT_LOGS_URL } from '@qovery/shared/routes'
import { queries } from '@qovery/state/util-queries'

export function useDeployAllServices() {
  const queryClient = useQueryClient()
  const navigate = useNavigate()

  return useMutation(mutations.deployAllServices, {
    onSuccess(_, { environment, payload }) {
      queryClient.invalidateQueries({
        queryKey: queries.services.listStatuses(environment.id).queryKey,
      })
      // NOTE: This is to invalidate deployed git_commit_id cache
      for (const { application_id: serviceId } of payload.applications ?? []) {
        queryClient.invalidateQueries({
          queryKey: queries.services.details({ serviceId, serviceType: 'APPLICATION' }).queryKey,
        })
      }
      for (const { id: serviceId } of payload.containers ?? []) {
        queryClient.invalidateQueries({
          queryKey: queries.services.details({ serviceId, serviceType: 'CONTAINER' }).queryKey,
        })
      }
      for (const id of payload.databases ?? []) {
        queryClient.invalidateQueries({
          queryKey: queries.services.details({ serviceId: id, serviceType: 'DATABASE' }).queryKey,
        })
      }
      for (const { id: serviceId } of payload.helms ?? []) {
        if (serviceId) {
          queryClient.invalidateQueries({
            queryKey: queries.services.details({ serviceId, serviceType: 'HELM' }).queryKey,
          })
        }
      }
      for (const { id: serviceId } of payload.jobs ?? []) {
        if (serviceId) {
          queryClient.invalidateQueries({
            queryKey: queries.services.details({ serviceId, serviceType: 'JOB' }).queryKey,
          })
        }
      }
    },
    meta: {
      notifyOnSuccess(_: unknown, variables: unknown) {
        const {
          environment: {
            id: environmentId,
            organization: { id: organizationId },
            project: { id: projectId },
          },
        } = variables as Parameters<typeof mutations.deployAllServices>[0]
        return {
          title: 'Your services are being updated',
          labelAction: 'See Deployment Logs',
          callback() {
            navigate(ENVIRONMENT_LOGS_URL(organizationId, projectId, environmentId))
          },
        }
      },
      notifyOnError: true,
    },
  })
}

export default useDeployAllServices
