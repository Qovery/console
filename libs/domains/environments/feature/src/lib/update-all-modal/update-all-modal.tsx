import { type Commit, type DeployAllRequest, type Environment } from 'qovery-typescript-axios'
import { useEffect, useState } from 'react'
import { P, match } from 'ts-pattern'
import { isJobGitSource } from '@qovery/shared/enums'
import {
  Button,
  Icon,
  InputCheckbox,
  LegacyAvatar,
  LegacyAvatarStyle,
  LoaderSpinner,
  ScrollShadowWrapper,
  TagCommit,
  Truncate,
  useModal,
} from '@qovery/shared/ui'
import { useDeployAllServices } from '../hooks/use-deploy-all-services/use-deploy-all-services'
import { type OutdatedService, useOutdatedServices } from '../hooks/use-outdated-services/use-outdated-services'

export interface UpdateAllModalProps {
  environment: Environment
}

export function UpdateAllModal({ environment }: UpdateAllModalProps) {
  const { closeModal } = useModal()
  const { data: outdatedServices = [], isLoading: isOutdatedServicesLoading } = useOutdatedServices({
    environmentId: environment.id,
  })
  const [selectedServiceIds, setSelectedServiceIds] = useState<string[]>([])
  const { mutate: deployAllServices, isLoading: isDeployAllServicesLoading } = useDeployAllServices()

  const checkService = (serviceId: string) => {
    if (selectedServiceIds.includes(serviceId)) {
      setSelectedServiceIds(selectedServiceIds.filter((id) => id !== serviceId))
    } else {
      setSelectedServiceIds([...selectedServiceIds, serviceId])
    }
  }

  const onSubmit = () => {
    if (selectedServiceIds.length > 0) {
      const appsToUpdate: DeployAllRequest['applications'] = []
      const jobsToUpdate: DeployAllRequest['jobs'] = []

      const servicesDictionary = outdatedServices.reduce(
        (acc, service) => {
          acc[service.id] = service
          return acc
        },
        {} as Record<string, OutdatedService>
      )

      selectedServiceIds.forEach((serviceId) => {
        const { serviceType, id, commits } = servicesDictionary[serviceId]
        if (serviceType === 'APPLICATION') {
          appsToUpdate.push({
            application_id: id,
            git_commit_id: commits[0]?.git_commit_id ?? '',
          })
        }

        if (serviceType === 'JOB') {
          jobsToUpdate.push({
            id,
            git_commit_id: commits[0]?.git_commit_id ?? '',
          })
        }
      })

      const deployRequest: DeployAllRequest = {
        applications: appsToUpdate,
        jobs: jobsToUpdate,
      }

      deployAllServices({ environment, payload: deployRequest })
      closeModal()
    }
  }

  const unselectAll = () => {
    setSelectedServiceIds([])
  }

  const selectAll = () => {
    setSelectedServiceIds(outdatedServices.map(({ id }) => id))
  }

  useEffect(() => {
    if (!isOutdatedServicesLoading) {
      selectAll()
    }
  }, [isOutdatedServicesLoading])

  const findCommitById = (application: OutdatedService, commitId?: string): Commit | undefined => {
    return application.commits.find((commit) => commit.git_commit_id === commitId) as Commit
  }

  const getAvatarForCommit = (application: OutdatedService, commitId?: string): string | undefined => {
    return findCommitById(application, commitId)?.author_avatar_url
  }

  const getNameForCommit = (application: OutdatedService, commitId?: string): string | undefined => {
    return findCommitById(application, commitId)?.author_name
  }
  const isChecked = (serviceId: string) => selectedServiceIds.includes(serviceId)

  return (
    <div className="p-6">
      <h2 className="h4 mb-1 max-w-sm truncate text-neutral-400 dark:text-neutral-50">Deploy latest version for..</h2>
      <p className="mb-4 text-sm text-neutral-350 dark:text-neutral-50">
        Select the services you want to update to the latest version
      </p>

      <div className="mb-4 flex items-center justify-between text-sm text-neutral-400 dark:text-neutral-50">
        <p>
          For{' '}
          <strong className="font-medium">
            <Truncate truncateLimit={60} text={environment?.name || ''} />
          </strong>
        </p>

        {outdatedServices.length > 0 &&
          (selectedServiceIds.length > 0 ? (
            <Button onClick={unselectAll} data-testid="deselect-all" size="sm" variant="surface" color="neutral">
              Deselect All
            </Button>
          ) : (
            <Button onClick={selectAll} data-testid="select-all" size="sm" variant="surface" color="neutral">
              Select All
            </Button>
          ))}
      </div>
      {isOutdatedServicesLoading ? (
        <LoaderSpinner className="mx-auto block" />
      ) : outdatedServices.length ? (
        <ScrollShadowWrapper className="max-h-[440px]">
          <ul>
            {outdatedServices.map((application, index) => {
              const gitRepository = match(application)
                .with({ serviceType: 'APPLICATION' }, ({ git_repository }) => git_repository)
                .with(
                  { serviceType: 'JOB', source: P.when(isJobGitSource) },
                  ({ source }) => source.docker?.git_repository
                )
                .otherwise(() => undefined)

              return (
                <li
                  data-testid="outdated-service-row"
                  onClick={() => checkService(application.id)}
                  key={application.id}
                  className={`${index === 0 ? 'rounded-t' : ''} ${
                    outdatedServices.length - 1 === index ? 'rounded-b !border-b' : ''
                  } flex justify-between border border-b-0 p-4 dark:border-neutral-400 ${
                    isChecked(application.id)
                      ? `border border-brand-500 bg-brand-50 dark:bg-neutral-500`
                      : 'border-neutral-250'
                  } ${outdatedServices && isChecked(outdatedServices[index - 1]?.id) && 'border-t-brand-500'}`}
                >
                  <div className="flex font-medium text-neutral-400 dark:text-neutral-50">
                    <InputCheckbox
                      name={application.id}
                      value={application.id}
                      isChecked={isChecked(application.id)}
                      className="mr-4"
                    />
                    <Truncate truncateLimit={31} text={application.name} />
                  </div>
                  <div className="ml-auto flex items-center">
                    <div
                      data-testid="current-commit-block"
                      className={`flex items-center ${isChecked(application.id) ? 'opacity-50' : ''}`}
                      onClick={(e) => e.stopPropagation()}
                    >
                      {}
                      <LegacyAvatar
                        size={28}
                        className="mr-2"
                        style={LegacyAvatarStyle.STROKED}
                        firstName={getNameForCommit(application, gitRepository?.deployed_commit_id) || 'Unknown'}
                        url={getAvatarForCommit(application, gitRepository?.deployed_commit_id)}
                      />
                      <TagCommit withBackground commitId={gitRepository?.deployed_commit_id} />
                    </div>
                    <Icon iconName="arrow-left" className="mx-2 -scale-100 text-neutral-400" />
                    {application.commits && Boolean(application.commits.length) && (
                      <div
                        data-testid="last-commit-block"
                        className={`flex items-center ${!isChecked(application.id) ? 'opacity-50' : ''}`}
                        onClick={(e) => e.stopPropagation()}
                      >
                        <LegacyAvatar
                          size={28}
                          className="mr-2"
                          style={LegacyAvatarStyle.STROKED}
                          firstName={application.commits[0].author_name || ''}
                          url={application.commits[0].author_avatar_url || ''}
                        />
                        <TagCommit withBackground commitId={application.commits[0].git_commit_id} />
                      </div>
                    )}
                  </div>
                </li>
              )
            })}
          </ul>
        </ScrollShadowWrapper>
      ) : (
        <div className="px-3 py-6 text-center" data-testid="empty-state">
          <Icon iconName="wave-pulse" className="text-neutral-350" />
          <p className="mt-1 text-xs font-medium text-neutral-350">No outdated services found</p>
        </div>
      )}

      <div className="sticky bottom-0 -mb-6 flex justify-end gap-3 bg-white py-6 dark:bg-neutral-550">
        <Button data-testid="cancel-button" color="neutral" variant="plain" size="lg" onClick={closeModal}>
          Cancel
        </Button>
        <Button
          data-testid="submit-button"
          disabled={selectedServiceIds.length === 0}
          type="submit"
          size="lg"
          onClick={onSubmit}
          loading={isDeployAllServicesLoading}
        >
          Update {selectedServiceIds.length} service{selectedServiceIds.length > 1 ? 's' : ''}
        </Button>
      </div>
    </div>
  )
}

export default UpdateAllModal
