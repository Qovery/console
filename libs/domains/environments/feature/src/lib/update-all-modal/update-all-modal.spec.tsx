import { type Commit, ServiceTypeEnum } from 'qovery-typescript-axios'
import { applicationFactoryMock, environmentFactoryMock } from '@qovery/shared/factories'
import { renderWithProviders, screen } from '@qovery/shared/util-tests'
import { useDeployAllServices } from '../hooks/use-deploy-all-services/use-deploy-all-services'
import { type OutdatedService } from '../hooks/use-outdated-services/use-outdated-services'
import { UpdateAllModal, type UpdateAllModalProps } from './update-all-modal'

const [mockEnvironment1, mockEnvironment2] = environmentFactoryMock(2).map((env, index) => ({
  ...env,
  id: `env${index + 1}`,
}))

const mockApplications = applicationFactoryMock(3).map((app): OutdatedService => {
  return {
    ...app,
    environment: {
      id: '1',
    },
    healthchecks: {},
    git_repository: {
      ...app.git_repository,
      url: '',
      deployed_commit_contributor: '',
      deployed_commit_date: '',
      deployed_commit_id: 'commit0',
      deployed_commit_tag: '',
    },
    serviceType: ServiceTypeEnum.APPLICATION,
    commits: [
      { git_commit_id: 'commit1' } as Commit,
      { git_commit_id: 'commit2' } as Commit,
      { git_commit_id: 'commit3' } as Commit,
    ],
  }
})

jest.mock('../hooks/use-outdated-services/use-outdated-services', () => ({
  useOutdatedServices: ({ environmentId }: { environmentId: string }) => ({
    data: environmentId === 'env1' ? mockApplications : [],
    isLoading: false,
    error: {},
  }),
}))

jest.mock('../hooks/use-deploy-all-services/use-deploy-all-services', () => {
  const mutate = jest.fn()
  return {
    ...jest.requireActual('../hooks/use-deploy-all-services/use-deploy-all-services'),
    useDeployAllServices: () => ({
      mutate,
      isLoading: false,
      error: {},
    }),
  }
})

describe('UpdateAllModal', () => {
  const props: UpdateAllModalProps = {
    environment: mockEnvironment1,
  }

  it('should render successfully', () => {
    const { baseElement } = renderWithProviders(<UpdateAllModal {...props} />)
    expect(baseElement).toBeTruthy()
  })

  it('should useDeployAllServices have been called on submit with good payload', async () => {
    const { userEvent } = renderWithProviders(<UpdateAllModal {...props} />)

    const deselectAll = screen.getByTestId('deselect-all')

    await userEvent.click(deselectAll)

    const firstRow = screen.getAllByTestId('outdated-service-row')[0]

    await userEvent.click(firstRow)

    const submitButton = screen.getByTestId('submit-button')

    await userEvent.click(submitButton)

    expect(useDeployAllServices().mutate).toHaveBeenCalledWith({
      environment: mockEnvironment1,
      payload: {
        applications: [
          {
            application_id: mockApplications[0].id,
            git_commit_id: mockApplications[0].commits[0].git_commit_id,
          },
        ],
        jobs: [],
      },
    })
  })

  it('should useDeployAllServices have been called on submit with selectAll', async () => {
    const { userEvent } = renderWithProviders(<UpdateAllModal {...props} />)

    const deselectAll = screen.getByTestId('deselect-all')

    await userEvent.click(deselectAll)

    await userEvent.click(screen.getByTestId('select-all'))

    const submitButton = screen.getByTestId('submit-button')

    await userEvent.click(submitButton)

    expect(useDeployAllServices().mutate).toHaveBeenCalledWith({
      environment: mockEnvironment1,
      payload: {
        applications: [
          {
            application_id: mockApplications[0].id,
            git_commit_id: mockApplications[0].commits[0].git_commit_id,
          },
          {
            application_id: mockApplications[1].id,
            git_commit_id: mockApplications[1].commits[0].git_commit_id,
          },
          {
            application_id: mockApplications[2].id,
            git_commit_id: mockApplications[2].commits[0].git_commit_id,
          },
        ],
        jobs: [],
      },
    })
  })

  it('should display 3 rows', () => {
    renderWithProviders(<UpdateAllModal {...props} />)
    expect(screen.getAllByTestId('outdated-service-row')).toHaveLength(3)
  })

  it('should have 1 row checked and next sibling color border top', async () => {
    const { userEvent } = renderWithProviders(<UpdateAllModal {...props} />)

    await userEvent.click(screen.getByTestId('deselect-all'))

    const rows = screen.getAllByTestId('outdated-service-row')
    const firstRow = rows[0]
    await userEvent.click(firstRow)

    expect(firstRow).toHaveClass('bg-brand-50', 'border-brand-500')

    const secondRow = rows[1]
    expect(secondRow).toHaveClass('border-t-brand-500')

    const submitButton = screen.getByTestId('submit-button')

    expect(submitButton).toHaveTextContent('Update 1 service')
  })

  it('should disabled submit button if no selected', async () => {
    const { userEvent } = renderWithProviders(<UpdateAllModal {...props} />)

    await userEvent.click(screen.getByTestId('deselect-all'))

    const submitButton = screen.getByTestId('submit-button')
    expect(submitButton).toHaveTextContent('Update 0 service')
    expect(submitButton).toBeDisabled()
  })

  it('should display empty state', async () => {
    renderWithProviders(<UpdateAllModal {...props} environment={mockEnvironment2} />)

    screen.getByTestId('empty-state')
  })

  it('should reduce opacity of commit blocks', async () => {
    const { baseElement, userEvent } = renderWithProviders(<UpdateAllModal {...props} />)

    await userEvent.click(screen.getByTestId('deselect-all'))

    const rows = screen.getAllByTestId('outdated-service-row')
    const firstRow = rows[0]
    await userEvent.click(firstRow)

    expect(baseElement).toMatchSnapshot()
  })
})
