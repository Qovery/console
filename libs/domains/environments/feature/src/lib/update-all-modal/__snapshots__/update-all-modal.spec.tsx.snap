// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateAllModal should reduce opacity of commit blocks 1`] = `
<body>
  <div>
    <div
      class="p-6"
    >
      <h2
        class="h4 mb-1 max-w-sm truncate text-neutral-400 dark:text-neutral-50"
      >
        Deploy latest version for..
      </h2>
      <p
        class="mb-4 text-sm text-neutral-350 dark:text-neutral-50"
      >
        Select the services you want to update to the latest version
      </p>
      <div
        class="mb-4 flex items-center justify-between text-sm text-neutral-400 dark:text-neutral-50"
      >
        <p>
          For
           
          <strong
            class="font-medium"
          >
            Barry Lyons
          </strong>
        </p>
        <button
          class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-brand-500 text-xs h-7 px-2 rounded border border-neutral-250 dark:border-neutral-400 text-neutral-400 dark:text-white hover:[&:not(:active)]:border-neutral-300 dark:hover:[&:not(:active):not(:disabled)]:border-neutral-350 active:bg-neutral-150 disabled:text-neutral-300 disabled:bg-neutral-100 disabled:border-none bg-neutral-100 dark:bg-neutral-500 hover:[&:not(:active):not(:disabled)]:bg-white dark:hover:[&:not(:active):not(:disabled)]:bg-neutral-400"
          data-testid="deselect-all"
        >
          Deselect All
        </button>
      </div>
      <div
        class="relative overflow-y-auto pr-[1px] max-h-[440px]"
        data-testid="scroll-shadow-wrapper"
      >
        <div
          class="pointer-events-none sticky top-0 -mb-4 h-4 w-full bg-scroll-shadow-up transition-opacity duration-300 opacity-0"
          data-testid="scroll-shadow-top"
        />
        <ul>
          <li
            class="rounded-t  flex justify-between border border-b-0 p-4 dark:border-neutral-400 border border-brand-500 bg-brand-50 dark:bg-neutral-500 false"
            data-testid="outdated-service-row"
          >
            <div
              class="flex font-medium text-neutral-400 dark:text-neutral-50"
            >
              <div
                class="relative flex items-center gap-2 mr-4"
              >
                <input
                  class="input-checkbox relative h-0 w-0 appearance-none font-icons before:absolute before:left-0 before:top-0 before:flex before:-translate-y-1/2 before:items-center before:justify-center before:rounded-sm before:bg-white before:text-white dark:before:bg-neutral-500 mr-5 before:w-4 before:h-4 cursor-pointer before:border-neutral-350 before:border-2 before:text-xs before:font-black before:leading-none before:transition-all before:content-[''] checked:before:border-brand-500 checked:before:bg-brand-500"
                  data-testid="input-checkbox"
                  id="0"
                  name="0"
                  type="checkbox"
                  value="0"
                />
                <i
                  aria-hidden="true"
                  class="fa-solid fa-check absolute text-2xs text-white left-1"
                />
              </div>
              Amy Maxwell
            </div>
            <div
              class="ml-auto flex items-center"
            >
              <div
                class="flex items-center opacity-50"
                data-testid="current-commit-block"
              >
                <div
                  class="relative block rounded-full border border-neutral-200  mr-2"
                  data-state="closed"
                  data-testid="avatar"
                  style="width: 28px; height: 28px;"
                >
                  <div
                    class="flex h-full w-full items-center justify-center rounded-full bg-neutral-200 text-center"
                  >
                    <span
                      class="relative text-xs font-medium text-neutral-400"
                    >
                      U
                    </span>
                  </div>
                </div>
                <div
                  class="block w-fit cursor-pointer"
                  data-state="closed"
                  data-testid="tag-commit"
                >
                  <span
                    class="inline-flex items-center rounded text-xs font-bold h-7 px-2   flex items-center justify-center border font-medium bg-white dark:bg-transparent w-[90px] border-neutral-250 text-neutral-350 hover:bg-neutral-200 dark:text-neutral-50 dark:hover:bg-neutral-400"
                    data-testid="tag"
                  >
                    <div
                      class="mr-1 w-4"
                    >
                      <i
                        aria-hidden="true"
                        class="fa-regular fa-code-commit mr-1"
                      />
                    </div>
                    commit0
                  </span>
                </div>
              </div>
              <i
                aria-hidden="true"
                class="fa-solid fa-arrow-left mx-2 -scale-100 text-neutral-400"
              />
              <div
                class="flex items-center "
                data-testid="last-commit-block"
              >
                <div
                  class="relative block rounded-full border border-neutral-200  mr-2"
                  data-state="closed"
                  data-testid="avatar"
                  style="width: 28px; height: 28px;"
                >
                  <div
                    class="flex h-full w-full items-center justify-center rounded-full bg-neutral-200 text-center"
                  >
                    <span
                      class="relative text-xs font-medium text-neutral-400"
                    />
                  </div>
                </div>
                <div
                  class="block w-fit cursor-pointer"
                  data-state="closed"
                  data-testid="tag-commit"
                >
                  <span
                    class="inline-flex items-center rounded text-xs font-bold h-7 px-2   flex items-center justify-center border font-medium bg-white dark:bg-transparent w-[90px] border-neutral-250 text-neutral-350 hover:bg-neutral-200 dark:text-neutral-50 dark:hover:bg-neutral-400"
                    data-testid="tag"
                  >
                    <div
                      class="mr-1 w-4"
                    >
                      <i
                        aria-hidden="true"
                        class="fa-regular fa-code-commit mr-1"
                      />
                    </div>
                    commit1
                  </span>
                </div>
              </div>
            </div>
          </li>
          <li
            class="  flex justify-between border border-b-0 p-4 dark:border-neutral-400 border-neutral-250 border-t-brand-500"
            data-testid="outdated-service-row"
          >
            <div
              class="flex font-medium text-neutral-400 dark:text-neutral-50"
            >
              <div
                class="relative flex items-center gap-2 mr-4"
              >
                <input
                  class="input-checkbox relative h-0 w-0 appearance-none font-icons before:absolute before:left-0 before:top-0 before:flex before:-translate-y-1/2 before:items-center before:justify-center before:rounded-sm before:bg-white before:text-white dark:before:bg-neutral-500 mr-5 before:w-4 before:h-4 cursor-pointer before:border-neutral-350 before:border-2 before:text-xs before:font-black before:leading-none before:transition-all before:content-[''] checked:before:border-brand-500 checked:before:bg-brand-500"
                  data-testid="input-checkbox"
                  id="1"
                  name="1"
                  type="checkbox"
                  value="1"
                />
              </div>
              Jeff Rhodes
            </div>
            <div
              class="ml-auto flex items-center"
            >
              <div
                class="flex items-center "
                data-testid="current-commit-block"
              >
                <div
                  class="relative block rounded-full border border-neutral-200  mr-2"
                  data-state="closed"
                  data-testid="avatar"
                  style="width: 28px; height: 28px;"
                >
                  <div
                    class="flex h-full w-full items-center justify-center rounded-full bg-neutral-200 text-center"
                  >
                    <span
                      class="relative text-xs font-medium text-neutral-400"
                    >
                      U
                    </span>
                  </div>
                </div>
                <div
                  class="block w-fit cursor-pointer"
                  data-state="closed"
                  data-testid="tag-commit"
                >
                  <span
                    class="inline-flex items-center rounded text-xs font-bold h-7 px-2   flex items-center justify-center border font-medium bg-white dark:bg-transparent w-[90px] border-neutral-250 text-neutral-350 hover:bg-neutral-200 dark:text-neutral-50 dark:hover:bg-neutral-400"
                    data-testid="tag"
                  >
                    <div
                      class="mr-1 w-4"
                    >
                      <i
                        aria-hidden="true"
                        class="fa-regular fa-code-commit mr-1"
                      />
                    </div>
                    commit0
                  </span>
                </div>
              </div>
              <i
                aria-hidden="true"
                class="fa-solid fa-arrow-left mx-2 -scale-100 text-neutral-400"
              />
              <div
                class="flex items-center opacity-50"
                data-testid="last-commit-block"
              >
                <div
                  class="relative block rounded-full border border-neutral-200  mr-2"
                  data-state="closed"
                  data-testid="avatar"
                  style="width: 28px; height: 28px;"
                >
                  <div
                    class="flex h-full w-full items-center justify-center rounded-full bg-neutral-200 text-center"
                  >
                    <span
                      class="relative text-xs font-medium text-neutral-400"
                    />
                  </div>
                </div>
                <div
                  class="block w-fit cursor-pointer"
                  data-state="closed"
                  data-testid="tag-commit"
                >
                  <span
                    class="inline-flex items-center rounded text-xs font-bold h-7 px-2   flex items-center justify-center border font-medium bg-white dark:bg-transparent w-[90px] border-neutral-250 text-neutral-350 hover:bg-neutral-200 dark:text-neutral-50 dark:hover:bg-neutral-400"
                    data-testid="tag"
                  >
                    <div
                      class="mr-1 w-4"
                    >
                      <i
                        aria-hidden="true"
                        class="fa-regular fa-code-commit mr-1"
                      />
                    </div>
                    commit1
                  </span>
                </div>
              </div>
            </div>
          </li>
          <li
            class=" rounded-b !border-b flex justify-between border border-b-0 p-4 dark:border-neutral-400 border-neutral-250 false"
            data-testid="outdated-service-row"
          >
            <div
              class="flex font-medium text-neutral-400 dark:text-neutral-50"
            >
              <div
                class="relative flex items-center gap-2 mr-4"
              >
                <input
                  class="input-checkbox relative h-0 w-0 appearance-none font-icons before:absolute before:left-0 before:top-0 before:flex before:-translate-y-1/2 before:items-center before:justify-center before:rounded-sm before:bg-white before:text-white dark:before:bg-neutral-500 mr-5 before:w-4 before:h-4 cursor-pointer before:border-neutral-350 before:border-2 before:text-xs before:font-black before:leading-none before:transition-all before:content-[''] checked:before:border-brand-500 checked:before:bg-brand-500"
                  data-testid="input-checkbox"
                  id="2"
                  name="2"
                  type="checkbox"
                  value="2"
                />
              </div>
              Joseph Morgan
            </div>
            <div
              class="ml-auto flex items-center"
            >
              <div
                class="flex items-center "
                data-testid="current-commit-block"
              >
                <div
                  class="relative block rounded-full border border-neutral-200  mr-2"
                  data-state="closed"
                  data-testid="avatar"
                  style="width: 28px; height: 28px;"
                >
                  <div
                    class="flex h-full w-full items-center justify-center rounded-full bg-neutral-200 text-center"
                  >
                    <span
                      class="relative text-xs font-medium text-neutral-400"
                    >
                      U
                    </span>
                  </div>
                </div>
                <div
                  class="block w-fit cursor-pointer"
                  data-state="closed"
                  data-testid="tag-commit"
                >
                  <span
                    class="inline-flex items-center rounded text-xs font-bold h-7 px-2   flex items-center justify-center border font-medium bg-white dark:bg-transparent w-[90px] border-neutral-250 text-neutral-350 hover:bg-neutral-200 dark:text-neutral-50 dark:hover:bg-neutral-400"
                    data-testid="tag"
                  >
                    <div
                      class="mr-1 w-4"
                    >
                      <i
                        aria-hidden="true"
                        class="fa-regular fa-code-commit mr-1"
                      />
                    </div>
                    commit0
                  </span>
                </div>
              </div>
              <i
                aria-hidden="true"
                class="fa-solid fa-arrow-left mx-2 -scale-100 text-neutral-400"
              />
              <div
                class="flex items-center opacity-50"
                data-testid="last-commit-block"
              >
                <div
                  class="relative block rounded-full border border-neutral-200  mr-2"
                  data-state="closed"
                  data-testid="avatar"
                  style="width: 28px; height: 28px;"
                >
                  <div
                    class="flex h-full w-full items-center justify-center rounded-full bg-neutral-200 text-center"
                  >
                    <span
                      class="relative text-xs font-medium text-neutral-400"
                    />
                  </div>
                </div>
                <div
                  class="block w-fit cursor-pointer"
                  data-state="closed"
                  data-testid="tag-commit"
                >
                  <span
                    class="inline-flex items-center rounded text-xs font-bold h-7 px-2   flex items-center justify-center border font-medium bg-white dark:bg-transparent w-[90px] border-neutral-250 text-neutral-350 hover:bg-neutral-200 dark:text-neutral-50 dark:hover:bg-neutral-400"
                    data-testid="tag"
                  >
                    <div
                      class="mr-1 w-4"
                    >
                      <i
                        aria-hidden="true"
                        class="fa-regular fa-code-commit mr-1"
                      />
                    </div>
                    commit1
                  </span>
                </div>
              </div>
            </div>
          </li>
        </ul>
        <div
          class="pointer-events-none sticky bottom-0 -mt-4 h-4 w-full rotate-180 bg-scroll-shadow-bottom transition-opacity duration-300 opacity-0"
          data-testid="scroll-shadow-bottom"
        />
      </div>
      <div
        class="sticky bottom-0 -mb-6 flex justify-end gap-3 bg-white py-6 dark:bg-neutral-550"
      >
        <button
          class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 hover:[&:not(:focus-visible):not(:disabled)]:shadow-sm active:[&:not(:focus-visible)]:shadow-none outline-brand-500 text-sm h-11 px-5 rounded border border-transparent text-neutral-350 hover:[&:not(:active):not(:focus-visible)]:border-neutral-300 hover:[&:not(:disabled)]:text-neutral-400 dark:hover:[&:not(:disabled)]:text-neutral-250 focus-visible:text-neutral-400 active:bg-neutral-150 dark:active:bg-transparent active:text-neutral-400 disabled:text-neutral-300 disabled:bg-neutral-150 disabled:border-none"
          data-testid="cancel-button"
        >
          Cancel
        </button>
        <button
          class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-neutral-500 text-sm h-11 px-5 rounded bg-brand-500 active:bg-brand-600 hover:bg-brand-400 text-white disabled:text-brand-300 disabled:bg-brand-100"
          data-testid="submit-button"
          type="submit"
        >
          Update 
          1
           service
        </button>
      </div>
    </div>
  </div>
</body>
`;
